File: backend-php/filters/AuthFilter.php
<?php
/**
 * Filtre CORS + authentification JWT.
 * Équivalent PHP de com.ihecride.filters.AuthFilter.
 *
 * Mode d'emploi :
 *   AuthFilter::cors();             // toujours, en tout début de requête
 *   AuthFilter::handlePreflight();  // court-circuit OPTIONS
 *   $ctx = AuthFilter::requireAuth();  // sur les routes protégées : retourne ['userId','role','email']
 */

require_once __DIR__ . '/../utils/JwtUtil.php';

class AuthFilter {

    /** Pose les en-têtes CORS sur la réponse. */
    public static function cors(): void {
        $origin = $_SERVER['HTTP_ORIGIN'] ?? '';
        if ($origin !== '') {
            header("Access-Control-Allow-Origin: $origin");
            header('Vary: Origin');
        } else {
            header('Access-Control-Allow-Origin: *');
        }
        header('Access-Control-Allow-Methods: GET, POST, PUT, DELETE, PATCH, OPTIONS');
        header('Access-Control-Allow-Headers: Authorization, Content-Type, Accept, X-Requested-With');
        header('Access-Control-Max-Age: 3600');
    }

    /** Si la méthode est OPTIONS, répond 204 et termine. */
    public static function handlePreflight(): void {
        if (($_SERVER['REQUEST_METHOD'] ?? '') === 'OPTIONS') {
            http_response_code(204);
            exit;
        }
    }

    /**
     * Sur une route protégée : valide le JWT, renvoie le contexte utilisateur.
     * En cas d'échec : 401 JSON et exit.
     * @return array{userId:int, role:string, email:string}
     */
    public static function requireAuth(): array {
        $auth = self::extractAuthHeader();
        if ($auth === null || strpos($auth, 'Bearer ') !== 0) {
            self::deny('Token manquant');
        }
        $token  = substr($auth, 7);
        $claims = JwtUtil::validate($token);
        if ($claims === null) {
            self::deny('Token invalide ou expiré');
        }
        return [
            'userId' => (int) $claims['sub'],
            'role'   => (string) ($claims['role'] ?? ''),
            'email'  => (string) ($claims['email'] ?? ''),
        ];
    }

    /** Renvoie 403 si le rôle ne correspond pas. */
    public static function requireRole(array $ctx, string $expectedRole): void {
        if (($ctx['role'] ?? '') !== $expectedRole) {
            http_response_code(403);
            header('Content-Type: application/json; charset=UTF-8');
            echo json_encode(['success' => false, 'message' => 'Accès refusé']);
            exit;
        }
    }

    private static function extractAuthHeader(): ?string {
        if (isset($_SERVER['HTTP_AUTHORIZATION'])) return $_SERVER['HTTP_AUTHORIZATION'];
        if (function_exists('apache_request_headers')) {
            $h = apache_request_headers();
            foreach ($h as $k => $v) {
                if (strtolower($k) === 'authorization') return $v;
            }
        }
        return $_SERVER['REDIRECT_HTTP_AUTHORIZATION'] ?? null;
    }

    private static function deny(string $msg): void {
        http_response_code(401);
        header('Content-Type: application/json; charset=UTF-8');
        echo json_encode(['success' => false, 'message' => $msg]);
        exit;
    }
}

File: backend/src/main/java/com/ihecride/filters/AuthFilter.java
package com.ihecride.filters;

import com.ihecride.utils.JwtUtil;
import io.jsonwebtoken.Claims;

import javax.servlet.*;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

/**
 * Filtre CORS + authentification JWT.
 *
 * Chemins publics (sans token) :
 *   /api/register, /api/login, /api/refresh
 *   /api/auth/register, /api/auth/login, /api/auth/refresh   (rétro-compat)
 *
 * CORS : autorise localhost (port 80, 8080, file://) en dev.
 * Sur OPTIONS, court-circuit immédiat avec 204 No Content.
 */
public class AuthFilter implements Filter {

    private static final Set<String> PUBLIC_PATHS = new HashSet<>(Arrays.asList(
        "/api/register",      "/api/login",      "/api/refresh",
        "/api/auth/register", "/api/auth/login", "/api/auth/refresh"
    ));

    @Override
    public void doFilter(ServletRequest req, ServletResponse res, FilterChain chain)
            throws IOException, ServletException {

        HttpServletRequest request   = (HttpServletRequest) req;
        HttpServletResponse response = (HttpServletResponse) res;

        // ---- En-têtes CORS sur toutes les réponses ----
        String origin = request.getHeader("Origin");
        if (origin != null && !origin.isEmpty()) {
            // Echo the Origin (works for http://localhost, http://localhost:80, file://, etc.)
            response.setHeader("Access-Control-Allow-Origin", origin);
            response.setHeader("Vary", "Origin");
        } else {
            response.setHeader("Access-Control-Allow-Origin", "*");
        }
        response.setHeader("Access-Control-Allow-Methods", "GET, POST, PUT, DELETE, PATCH, OPTIONS");
        response.setHeader("Access-Control-Allow-Headers", "Authorization, Content-Type, Accept, X-Requested-With");
        response.setHeader("Access-Control-Max-Age", "3600");

        // ---- Preflight CORS : court-circuit ----
        if ("OPTIONS".equalsIgnoreCase(request.getMethod())) {
            response.setStatus(HttpServletResponse.SC_NO_CONTENT);
            return;
        }

        // ---- Routes publiques ----
        String path = request.getRequestURI();
        String ctx  = request.getContextPath();
        String relative = (ctx != null && path.startsWith(ctx)) ? path.substring(ctx.length()) : path;
        if (PUBLIC_PATHS.contains(relative)) {
            chain.doFilter(req, res);
            return;
        }

        // ---- Authentification JWT ----
        String auth = request.getHeader("Authorization");
        if (auth == null || !auth.startsWith("Bearer ")) {
            response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
            response.setContentType("application/json; charset=UTF-8");
            response.getWriter().write("{\"success\":false,\"message\":\"Token manquant\"}");
            return;
        }

        String token = auth.substring(7);
        Claims claims = JwtUtil.validate(token);
        if (claims == null) {
            response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
            response.setContentType("application/json; charset=UTF-8");
            response.getWriter().write("{\"success\":false,\"message\":\"Token invalide ou expiré\"}");
            return;
        }

        request.setAttribute("userId", Integer.parseInt(claims.getSubject()));
        request.setAttribute("role",   claims.get("role"));
        request.setAttribute("email",  claims.get("email"));
        chain.doFilter(req, res);
    }
}
